File: VC_HIL/python/inverterhilgui/validate_command_value.py
"""Fail-closed command type/range validation."""

import math
from .parameter_contract import contractEntry


def _scalar_number(value):
    return isinstance(value, (int, float, bool)) and not isinstance(value, complex) and math.isfinite(float(value))


def validateCommandValue(source, name, value):
    result = {"name": "", "value": None, "accepted": False, "clamped": False, "reason": "unknown_logical_name"}
    if not isinstance(name, str) or not name:
        result["reason"] = "malformed_logical_name"
        return result
    result["name"] = name
    entry = contractEntry(source, name)
    if entry is None:
        return result
    for field in ("type", "minimum", "maximum"):
        if not hasattr(entry, field):
            result["reason"] = f"missing_contract_{field}"
            return result
    if not all(_scalar_number(x) for x in (entry.minimum, entry.maximum)) or entry.minimum > entry.maximum:
        result["reason"] = "malformed_contract_range"
        return result
    if not _scalar_number(value):
        result["reason"] = "malformed_value"
        return result
    numeric = float(value)
    if entry.type == "double":
        output = min(max(numeric, entry.minimum), entry.maximum)
        result["clamped"] = output != numeric
        result["value"] = output
    elif entry.type == "logical":
        if numeric not in (0, 1):
            result["reason"] = "value_not_logical"
            return result
        result["value"] = bool(numeric)
    elif entry.type in ("uint8", "uint16", "uint32"):
        if numeric != math.floor(numeric):
            result["reason"] = "value_not_integer"
            return result
        if numeric < 0:
            result["reason"] = "value_negative"
            return result
        output = int(min(max(numeric, entry.minimum), entry.maximum))
        result["clamped"] = output != numeric
        result["value"] = output
    else:
        result["reason"] = "unsupported_contract_type"
        return result
    result["accepted"] = True
    result["reason"] = "clamped" if result["clamped"] else "accepted"
    return result

File: VC_HIL/inverter/+inverterhilgui/+writes/validateCommandValue.m
function result = validateCommandValue(source, name, value)
%VALIDATECOMMANDVALUE Validate and clamp one logical command before writing.
%
%   RESULT = VALIDATECOMMANDVALUE(SOURCE, NAME, VALUE) checks VALUE against
%   the declared type and range for logical NAME. SOURCE is either the
%   declarative contract or the resolved contract. This evaluator FAILS
%   CLOSED: a malformed value, an unknown name, or a malformed contract entry
%   returns ACCEPTED false with a REASON string instead of throwing, so no
%   callback can push an unvalidated value at the target.
%
%   RESULT fields:
%     name     - the requested logical name
%     value    - the value that may be written when ACCEPTED is true
%     accepted - true only when the value is safe to write
%     clamped  - true when VALUE was outside the declared range
%     reason   - 'accepted', 'clamped' or a specific rejection reason

result = struct( ...
    'name', '', ...
    'value', [], ...
    'accepted', false, ...
    'clamped', false, ...
    'reason', 'unknown_logical_name');
if isstring(name) && isscalar(name)
    name = char(name);
end
if ~ischar(name) || isempty(name)
    result.reason = 'malformed_logical_name';
    return;
end
result.name = name;

entry = inverterhilgui.params.contractEntry(source, name);
if isempty(entry)
    return;
end
required = {'type', 'minimum', 'maximum'};
for index = 1:numel(required)
    if ~isfield(entry, required{index})
        result.reason = ['missing_contract_' required{index}];
        return;
    end
end
if ~isnumeric(entry.minimum) || ~isscalar(entry.minimum) || ...
        ~isreal(entry.minimum) || ~isfinite(entry.minimum) || ...
        ~isnumeric(entry.maximum) || ~isscalar(entry.maximum) || ...
        ~isreal(entry.maximum) || ~isfinite(entry.maximum) || ...
        entry.minimum > entry.maximum
    result.reason = 'malformed_contract_range';
    return;
end

if ~(isnumeric(value) || islogical(value)) || ~isscalar(value) || ...
        ~isreal(value) || ~isfinite(double(value))
    result.reason = 'malformed_value';
    return;
end
numericValue = double(value);

switch entry.type
    case 'double'
        clamped = min(max(numericValue, entry.minimum), entry.maximum);
        result.clamped = clamped ~= numericValue;
        result.value = clamped;
    case 'logical'
        if numericValue ~= 0 && numericValue ~= 1
            result.reason = 'value_not_logical';
            return;
        end
        result.value = logical(numericValue);
    case {'uint8', 'uint16', 'uint32'}
        if numericValue ~= floor(numericValue)
            result.reason = 'value_not_integer';
            return;
        end
        if numericValue < 0
            result.reason = 'value_negative';
            return;
        end
        clamped = min(max(numericValue, entry.minimum), entry.maximum);
        result.clamped = clamped ~= numericValue;
        result.value = cast(clamped, entry.type);
    otherwise
        result.reason = 'unsupported_contract_type';
        return;
end

result.accepted = true;
if result.clamped
    result.reason = 'clamped';
else
    result.reason = 'accepted';
end
end
